src/__flow-tests__/generics-tests.js

```diff
@@ -6,53 +6,218 @@ import type { IsSupertype } from '../types';
+  //   $FlowExpectedError[incompatible-type-arg]
+  (x: IsString<mixed>): string => x;
+  // We could flow it into somewhere else instead, and still get the error:
+  //   $FlowExpectedError[incompatible-type-arg]
+  (x: IsString<mixed>): string => { (x: string); return 'b'; };
+  // But if it flows nowhere at all, there's no error:
+  (x: IsString<mixed>): string => { x; return 'b'; };
+  (x: IsString<mixed>): string => 'b';
+
+  // Similarly if we just mention the bogus type in another way but with no
+  // value-level flows to or from it, there's no error:
+  declare var x: IsString<number>;
+
+  //
+  //
+  // Happily this point isn't really a warning -- there's no need to worry
+  // about getting anything wrong by forgetting it.  As long as you set out
+  // to write tests for the error case, if you forget this point then you'll
+  // see that your tests are failing to fail.
+  //
+  // Rather, this is a tip for how to solve that problem if you run into it:
+  // cause some value to flow to or from the bogus type, and then Flow
+  // should notice that the type is bogus.
+}
 
+/** A subtler tip on writing tests for fancy generic types like these. */
+function demo_short_circuits() {
+  // As seen above, in order to test one of these types you need to exercise
+  // it with some value-level flow to or from it.
+  //
+  // A further wrinkle is that if your value-level flow is necessarily OK
+  // thanks to some general rule of the type system, some reasoning that
+  // doesn't depend on the details of the bogus type (that you're trying to
+  // test is indeed seen as bogus)… then Flow may apply that rule and
+  // short-circuit any closer inspection of the specific types on each side.
+  //
+  // In particular (and we'll demo each of these below):
+  //  * If the flow is from `empty`, it succeeds.  (Even if the other end is
+  //    an internally-bogus type.)
+  //  * If the flow is to `mixed`, it succeeds.  (Ditto.)
+  //  * If the flow is from some type to itself, it succeeds (even if that
+  //    type is internally bogus.)
+  //
+  // (The jargon for these facts is: `empty` is a "bottom" type; `mixed` is
+  // a "top" type; and the subtype relation (the ability to flow from one
+  // type to another) is "reflexive".)
+  //
+  // From Flow's perspective this is a fundamentally sound thing to do,
+  // because in the context of the whole program, either there is a type
+  // error somewhere else, or it is impossible for an actual value at
+  // runtime to traverse that data flow.  In particular:
+  //  * The `empty` type, as the name suggests, has no values.  So if an
+  //    expression has type `empty` and you ever reach that expression at
+  //    runtime, then there must have been a type error along the way.
+  //  * If the flow is from a bogus type to `mixed` or to itself, then that
+  //    flow can only be realized at runtime if a value somehow got into the
+  //    bogus type in the first place.
```

Comment:
> What if the flow is from a bogus type to something other than `mixed` or itself?

Then that's the normal case of using a bogus type. (Well, that together with a flow to a bogus type from something other than `empty` or itself.)

In that case (unless some other general rule of the type system applies, in the same spirit as the `empty <: T`, `T <: mixed`, and `T <: T` rules demo'd here), the flow should already cause Flow to look closer at the type and see it's bogus, and give an error.

> Does that imply that the flow might be realizable at runtime even if a value didn't get into the bogus type in the first place?

Hmm. The background idea here is meant to be: if there's a flow from point A in the program (some expression, or variable, or object property, etc.) to point B, then for an actual value at runtime to go from point A to point B, the value must have gotten into point A in the first place.

So in particular if there's a flow from point A which has some bogus type, to point B which has whatever type, and that flow is realized at runtime, then the value must have gotten to point A, which has that bogus type. So if there isn't a type error somewhere, then whatever flow got it *to* point A must have somehow not caused a type error.


> * When Flow checks a flow from `F` to `T`, it'll check as valid iff it thinks `F <: T`

Yeah. That's basically what the notation `<:` means when talking about Flow -- it means that Flow would accept a flow from something of type F to something of type T.

> (where `F` and `T` aren't unclear types like `any` or `Object`, I guess).

(The way I'd describe this is actually that for all T, `T <: any` and `any <: T`. Similarly `T <: Object` and `Object <: T` for any T that's an object type.)


> * If `T` is `mixed`, the flow is always checked as valid. There, Flow reasonably thinks `F <: T` because `mixed` is a "top" type. And `mixed` is a "top" type whether or not `F` is bogus, so Flow doesn't bother checking if `F` is bogus.
> * If `T` is `F`, the flow is always checked as valid. There, Flow reasonably thinks `F <: T` because the subtype relation is reflexive. And the subtype relation is reflexive whether or not `F` is bogus, so Flow doesn't bother checking if `F` is bogus.

Yep!


> * […] it sometimes doesn't catch that it can't be realized at runtime.

Hmm, perhaps this phrasing was confusing!

I mean "can't be realized at runtime" as a description of a *good* thing, a mitigating thing, when one of the ends of the flow is bogus.

That is: it seems bad that Flow cheerfully lets you write these bogus types. And it is somewhat bad and I think ideally Flow wouldn't do it. (One likely reason it doesn't is performance: it notices that some flow it's checking goes to `mixed`, and can immediately check it off as good and move on, without going and digging each time into the details of the source type.) But it's not *too* bad, because at least when that happens, it can't happen in a situation that's actually some part of the code that can actually operate at runtime (without a type error somewhere else.)

Does that help?

Reply:
> Does that help?

Yes, thanks very much!! 🙂